lib/base/logger.cpp

```diff
@@ -243,17 +244,35 @@ void Logger::UpdateMinLogSeverity()
 	m_MinLogSeverity.store(result);
 }
 
-Log::Log(LogSeverity severity, String facility, const String& message)
-	: Log(severity, std::move(facility))
+Log::Log(LogSeverity severity, String facility, const String& message, const std::vector<ConfigObject::Ptr>& involved)
+	: Log(severity, std::move(facility), involved)
 {
 	if (!m_IsNoOp) {
 		m_Buffer << message;
 	}
 }
 
-Log::Log(LogSeverity severity, String facility)
-	: m_Severity(severity), m_Facility(std::move(facility)), m_IsNoOp(severity < Logger::GetMinLogSeverity())
-{ }
+Log::Log(LogSeverity severity, String facility, const std::vector<ConfigObject::Ptr>& involved)
+	: m_Severity(severity), m_Facility(std::move(facility))
+{
+	std::set<ConfigObject::Ptr> objects;
+
+	for (auto& obj : involved) {
+		objects.merge(DependencyGraph::GetAllParents(obj));
+	}
+
+	objects.insert(involved.begin(), involved.end());
+
+	for (auto& obj : objects) {
+		auto severity (obj->GetOverrideSeverityRaw());
+
+		if (severity >= 0 && severity < m_OverriddenMinSeverity) {
+			m_OverriddenMinSeverity = (LogSeverity)severity;
+		}
+	}
+
+	m_IsNoOp = severity < std::min(Logger::GetMinLogSeverity(), m_OverriddenMinSeverity);
+}
```

Comment:
This looks really expensive to do for every single debug log message that's probably going to be discarded in the end.

What I think could work better: pass only a single object to `Log()` and only compare a single attribute of that object. That attribute would store the combined log severity of that object, i.e. the minimal severity configured on that object or any of its parents. It would be updated when the severity is set, not used, i.e. put the runtime overhead there instead of when log messages are generated. That would also mean that there would only be runtime overhead if that feature is actually used.

(FYI: I haven't put much time into thinking about whether `DependencyGraph` actually stores a relation that's useful for this use-case or if it might include too much or too little.)